go/vt/vttablet/tabletserver/vstreamer/vstreamer.go

```diff
@@ -1055,6 +1085,111 @@ func (vs *vstreamer) extractRowAndFilter(plan *streamerPlan, data []byte, dataCo
+		}
+	}
+	return nil
+}
+
+// buildEnumtringValue takes the integer value of an ENUM column and returns the string value.
+func buildEnumStringValue(plan *streamerPlan, colNum int, value sqltypes.Value) (sqltypes.Value, error) {
+	if value.IsNull() { // No work is needed
+		return value, nil
+	}
+	// Add the mappings just-in-time in case we haven't properly received and processed a
+	// table map event to initialize it.
+	if plan.EnumSetValuesMap == nil {
+		if err := addEnumAndSetMappingstoPlan(plan.Plan, plan.Table.Fields, plan.TableMap.Metadata); err != nil {
+			return sqltypes.Value{}, err
+		}
+	}
+	// ENUM columns are stored as an unsigned 16-bit integer as they can contain a maximum
+	// of 65,535 elements (https://dev.mysql.com/doc/refman/en/enum.html) with the 0 element
+	// reserved for any integer value that has no string mapping.
+	iv, err := value.ToUint16()
+	if err != nil {
+		return sqltypes.Value{}, fmt.Errorf("no valid integer value found for column %s in table %s, bytes: %b",
+			plan.Table.Fields[colNum].Name, plan.Table.Name, iv)
+	}
+	var strVal string
+	// Match the MySQL behavior of returning an empty string for invalid ENUM values.
+	// This is what the 0 position in an ENUM is reserved for.
+	if iv != 0 {
+		var ok bool
+		strVal, ok = plan.EnumSetValuesMap[colNum][int(iv)]
+		if !ok {
+			// The integer value was NOT 0 yet we found no mapping. This should never happen.
+			return sqltypes.Value{}, fmt.Errorf("no string value found for ENUM column %s in table %s -- with available values being: %v -- using the found integer value: %d",
+				plan.Table.Fields[colNum].Name, plan.Table.Name, plan.EnumSetValuesMap[colNum], iv)
+		}
+	}
+	return sqltypes.MakeTrusted(plan.Table.Fields[colNum].Type, []byte(strVal)), nil
+}
+
+// buildSetStringValue takes the integer value of a SET column and returns the string value.
+func buildSetStringValue(plan *streamerPlan, colNum int, value sqltypes.Value) (sqltypes.Value, error) {
+	if value.IsNull() { // No work is needed
+		return value, nil
+	}
+	// Add the mappings just-in-time in case we haven't properly received and processed a
+	// table map event to initialize it.
+	if plan.EnumSetValuesMap == nil {
+		if err := addEnumAndSetMappingstoPlan(plan.Plan, plan.Table.Fields, plan.TableMap.Metadata); err != nil {
+			return sqltypes.Value{}, err
+		}
+	}
+	// A SET column can have 64 unique values: https://dev.mysql.com/doc/refman/en/set.html
+	// For this reason the binlog event contains the values encoded as an unsigned 64-bit
+	// integer which is really a bitmap (note that position 0 is reserved for '' which is
+	// used if you insert any integer values which have no valid string mapping in the set).
+	val := bytes.Buffer{}
+	iv, err := value.ToUint64()
+	if err != nil {
+		return value, fmt.Errorf("no valid integer value found for column %s in table %s, bytes: %b",
+			plan.Table.Fields[colNum].Name, plan.Table.Name, iv)
+	} else {
```

Comment:
Since you `return` in the above, please remove this `else` statement and unindent.

Reply:
Good catch. Thanks!